### ISSUE TITLE
Argument #1 ($element) must be of type Phpactor\DocblockParser\Ast\Element, null given at DocblockParser/Ast/TypeList.php#40

### ISSUE THREAD

**Issue by axelrindle** [phpactor/phpactor#2693]
## Description

Imagine the following model in Laravel:

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Test extends Model
{
    use HasFactory;

    public function test()
    {
        return $this->hasOne(User::class);
    }
}
```

Just opening the class (in VSCode) produces the following error on-mass:

```
[Error - 12:36:25] Request textDocument/codeAction failed.
  Message: Exception [TypeError] Phpactor\DocblockParser\Ast\TypeList::Phpactor\DocblockParser\Ast\{closure}(): Argument #1 ($element) must be of type Phpactor\DocblockParser\Ast\Element, null given at phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/DocblockParser/Ast/TypeList.php#40
  Code: -32603 
#0 [internal function]: Phpactor\DocblockParser\Ast\TypeList->Phpactor\DocblockParser\Ast\{closure}()
#1 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/DocblockParser/Ast/TypeList.php(40): array_filter()
#2 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php(274): Phpactor\DocblockParser\Ast\TypeList->types()
#3 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php(101): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\TypeConverter->convertGeneric()
#4 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/Phpactor/DocblockParser/ParsedDocblock.php(327): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\TypeConverter->convert()
#5 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/Phpactor/DocblockParser/ParsedDocblock.php(112): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\ParsedDocblock->convertType()
#6 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Reflection/TypeResolver/PropertyTypeResolver.php(41): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\ParsedDocblock->vars()
#7 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Reflection/TypeResolver/PropertyTypeResolver.php(17): Phpactor\WorseReflection\Core\Reflection\TypeResolver\PropertyTypeResolver->getDocblockType()
#8 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/TolerantParser/Reflection/ReflectionProperty.php(75): Phpactor\WorseReflection\Core\Reflection\TypeResolver\PropertyTypeResolver->resolve()
#9 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(198): Phpactor\WorseReflection\Bridge\TolerantParser\Reflection\ReflectionProperty->inferredType()
#10 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(95): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->addClassContext()
#11 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(55): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->walkFunctionLike()
#12 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/FrameResolver.php(141): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->enter()
#13 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#14 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#15 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#16 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/WorseReflection/Bridge/TolerantParser/Reflector/TolerantSourceCodeReflector.php(76): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#17 [internal function]: Phpactor\WorseReflection\Bridge\TolerantParser\Reflector\TolerantSourceCodeReflector->Phpactor\WorseReflection\Bridge\TolerantParser\Reflector\{closure}()
#18 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Coroutine.php(118): Generator->send()
#19 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#20 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Delayed.php(26): Amp\Delayed->resolve()
#21 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Loop/NativeDriver.php(142): Amp\Delayed->Amp\{closure}()
#22 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Loop/Driver.php(138): Amp\Loop\NativeDriver->dispatch()
#23 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#24 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#25 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(114): Amp\Loop::run()
#26 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/Extension/LanguageServer/Command/StartCommand.php(50): Phpactor\LanguageServer\Core\Server\LanguageServer->run()
#27 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/symfony/console/Command/Command.php(298): Phpactor\Extension\LanguageServer\Command\StartCommand->execute()
#28 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run()
#29 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand()
#30 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/lib/Application.php(48): Symfony\Component\Console\Application->doRun()
#31 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/vendor/symfony/console/Application.php(171): Phpactor\Application->doRun()
#32 phar:///home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar/bin/phpactor(45): Symfony\Component\Console\Application->run()
#33 /home/axel/.vscode/extensions/phpactor.vscode-phpactor-1.4.0/phpactor.phar(15): require('...')
#34 {main}
```

## Versions

phpactor: 2024.06.30.0
php: 8.3.9
Laravel: 11.15.0

**Comment by trizz** [comment]
Can confirm this also happens in my neovim setup.

**Comment by dantleech** [comment]
I've pushed a fix to master can you confirm it works?

**Comment by axelrindle** [comment]
That gives the following error:

```
Starting language server, use -vvv for verbose output
[Error - 12:55:02] Request textDocument/codeAction failed.
  Message: Exception [RuntimeException] List has no first element at /home/axel/Projekte/github.com/phpactor/lib/DocblockParser/Ast/TypeNodes.php#36
  Code: -32603 
#0 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php(274): Phpactor\DocblockParser\Ast\TypeNodes->first()
#1 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php(101): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\TypeConverter->convertGeneric()
#2 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/Phpactor/DocblockParser/ParsedDocblock.php(327): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\TypeConverter->convert()
#3 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/Phpactor/DocblockParser/ParsedDocblock.php(112): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\ParsedDocblock->convertType()
#4 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Reflection/TypeResolver/PropertyTypeResolver.php(41): Phpactor\WorseReflection\Bridge\Phpactor\DocblockParser\ParsedDocblock->vars()
#5 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Reflection/TypeResolver/PropertyTypeResolver.php(17): Phpactor\WorseReflection\Core\Reflection\TypeResolver\PropertyTypeResolver->getDocblockType()
#6 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/TolerantParser/Reflection/ReflectionProperty.php(75): Phpactor\WorseReflection\Core\Reflection\TypeResolver\PropertyTypeResolver->resolve()
#7 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(198): Phpactor\WorseReflection\Bridge\TolerantParser\Reflection\ReflectionProperty->inferredType()
#8 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(95): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->addClassContext()
#9 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/Walker/FunctionLikeWalker.php(55): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->walkFunctionLike()
#10 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/FrameResolver.php(141): Phpactor\WorseReflection\Core\Inference\Walker\FunctionLikeWalker->enter()
#11 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#12 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#13 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Core/Inference/FrameResolver.php(147): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#14 /home/axel/Projekte/github.com/phpactor/lib/WorseReflection/Bridge/TolerantParser/Reflector/TolerantSourceCodeReflector.php(76): Phpactor\WorseReflection\Core\Inference\FrameResolver->walkNode()
#15 [internal function]: Phpactor\WorseReflection\Bridge\TolerantParser\Reflector\TolerantSourceCodeReflector->Phpactor\WorseReflection\Bridge\TolerantParser\Reflector\{closure}()
#16 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Coroutine.php(118): Generator->send()
#17 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#18 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Delayed.php(26): Amp\Delayed->resolve()
#19 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Loop/NativeDriver.php(142): Amp\Delayed->Amp\{closure}()
#20 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(138): Amp\Loop\NativeDriver->dispatch()
#21 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#22 /home/axel/Projekte/github.com/phpactor/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#23 /home/axel/Projekte/github.com/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(116): Amp\Loop::run()
#24 /home/axel/Projekte/github.com/phpactor/lib/Extension/LanguageServer/Command/StartCommand.php(50): Phpactor\LanguageServer\Core\Server\LanguageServer->run()
#25 /home/axel/Projekte/github.com/phpactor/vendor/symfony/console/Command/Command.php(298): Phpactor\Extension\LanguageServer\Command\StartCommand->execute()
#26 /home/axel/Projekte/github.com/phpactor/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run()
#27 /home/axel/Projekte/github.com/phpactor/vendor/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand()
#28 /home/axel/Projekte/github.com/phpactor/lib/Application.php(48): Symfony\Component\Console\Application->doRun()
#29 /home/axel/Projekte/github.com/phpactor/vendor/symfony/console/Application.php(171): Phpactor\Application->doRun()
#30 /home/axel/Projekte/github.com/phpactor/bin/phpactor(46): Symfony\Component\Console\Application->run()
#31 {main}
```

**Comment by dantleech** [comment]
Ideally it would be good to get the minimum code required to reproduce this error and a test case... or:

```diff
diff --git a/lib/DocblockParser/Ast/TypeNodes.php b/lib/DocblockParser/Ast/TypeNodes.php
index 080dc15b..b6d8be15 100644
--- a/lib/DocblockParser/Ast/TypeNodes.php
+++ b/lib/DocblockParser/Ast/TypeNodes.php
@@ -3,6 +3,7 @@
 namespace Phpactor\DocblockParser\Ast;
 
 use ArrayIterator;
+use Countable;
 use IteratorAggregate;
 use RuntimeException;
 use Traversable;
@@ -10,7 +11,7 @@ use Traversable;
 /**
  * @implements IteratorAggregate<TypeNode>
  */
-class TypeNodes implements IteratorAggregate
+class TypeNodes implements IteratorAggregate, Countable
 {
     /**
      * @var TypeNode[]
@@ -37,4 +38,9 @@ class TypeNodes implements IteratorAggregate
             'List has no first element'
         ));
     }
+
+    public function count(): int
+    {
+        return count($this->types);
+    }
 }
diff --git a/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php b/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php
index 29a81f5c..28e734c6 100644
--- a/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php
+++ b/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php
@@ -268,7 +268,7 @@ class TypeConverter
 
         if ($classType instanceof ClassStringType) {
             $parameters = $type->parameters();
-            if ($parameters->count()) {
+            if ($parameters->types()->count()) {
                 return new ClassStringType(
                     ClassName::fromString($this->convert(
                         $parameters->types()->first()
```

if  you apply that diff it should work, then I'll ... push to master.

**Comment by axelrindle** [comment]
Seems to work. Thank you! 🎉

**Comment by dantleech** [comment]
cool, pushed to master. _I should really protect that branch from myself_.

**Comment by Illizian** [comment]
This is working for me too. Is it possible to get this released, as I'm unable to manually install this on my work machine.

**Comment by adaviloper** [comment]
Sorry in advance since I'm unaware of the normal process, but is this going to be shipped in a new patch? The latest release is `2024.06.30.0`.

**Comment by dantleech** [comment]
There's no regular process unfortunately, but I've tagged a new release https://github.com/phpactor/phpactor/releases/tag/2024.11.05.0
